src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -390,6 +390,82 @@ def test_all_endpoints_have_valid_composition(self):
         asserts.skip("TODO: Make a test that verifies each endpoint only lists direct descendants, except Root Node and Aggregator endpoints that list all their descendants")
 
+    def test_TC_PS_3_1(self):
+        success = True
+        self.print_step(1, "Wildcard read of device - already done")
+
+        self.print_step(2, "Verify that all endpoints listed in the EndpointList are valid")
+        attribute_id = Clusters.PowerSource.Attributes.EndpointList.attribute_id
+        cluster_id = Clusters.PowerSource.id
+        attribute_string = self.cluster_mapper.get_attribute_string(cluster_id, attribute_id)
+        for endpoint_id, endpoint in self.endpoints.items():
+            if Clusters.PowerSource not in endpoint:
+                continue
+            if Clusters.PowerSource.Attributes.EndpointList not in endpoint[Clusters.PowerSource]:
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem=f'Did not find {attribute_string} on {location.as_cluster_string(self.cluster_mapper)}', spec_location="EndpointList Attribute")
+                success = False
+            endpoint_list = endpoint[Clusters.PowerSource][Clusters.PowerSource.Attributes.EndpointList]
+            if endpoint_list and not all(id in self.endpoints.keys() for id in endpoint_list):
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem=f'{attribute_string} lists a non-existent endpoint', spec_location="EndpointList Attribute")
+                success = False
+
+        self.print_step(3, "Verify that all Bridged Node endpoint lists are correct")
+        device_types = {}
+        parts_list = {}
+        for endpoint_id, endpoint in self.endpoints.items():
+            if Clusters.PowerSource not in endpoint or Clusters.PowerSource.Attributes.EndpointList not in endpoint[Clusters.PowerSource]:
+                continue
+            if Clusters.Descriptor not in endpoint or Clusters.Descriptor.Attributes.PartsList not in endpoint[Clusters.Descriptor] or Clusters.Descriptor.Attributes.DeviceTypeList not in endpoint[Clusters.Descriptor]:
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=Clusters.Descriptor.id,
+                                                 attribute_id=Clusters.Descriptor.Attributes.PartsList.id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem='Missing descriptor cluster PartsList or DeviceTypeList', spec_location="PartsList Attribute")
+                success = False
+            device_types[endpoint_id] = [i.deviceType for i in endpoint[Clusters.Descriptor]
```

Comment:
If the above fails, these may not be valid. Is the default always an empty list? What if Clusters.Descriptor is not on the endpoint, would the map operator work?

Reply:
just added a continue, so we skip this.